sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -330,4 +362,139 @@ object FileStreamSource {
 
     def size: Int = map.size()
   }
+
+  private[sql] class FileStreamSourceCleaner(
+      fileSystem: FileSystem,
+      sourcePath: Path,
+      baseArchiveFileSystem: Option[FileSystem],
+      baseArchivePath: Option[Path]) extends Logging {
+    require(baseArchiveFileSystem.isDefined == baseArchivePath.isDefined)
+
+    private val sourceGlobFilters: Seq[GlobFilter] = buildSourceGlobFilters(sourcePath)
+
+    private val sameFsSourceAndArchive: Boolean = {
+      baseArchiveFileSystem.exists { fs =>
+        if (fileSystem.getUri != fs.getUri) {
+          logWarning("Base archive path is located to the different filesystem with source, " +
+            s"which is not supported. source path: ${sourcePath} / base archive path: " +
+            s"${baseArchivePath.get}")
+          false
+        } else {
+          true
+        }
+      }
+    }
+
+    /**
+     * This is a flag to skip matching archived path with source path.
+     *
+     * FileStreamSource reads the files which one of below conditions is met:
+     * 1) file itself is matched with source path
+     * 2) parent directory is matched with source path
+     *
+     * Checking with glob pattern is costly, so this flag leverages above information to prune
+     * the cases where the file cannot be matched with source path. For example, when file is
+     * moved to archive directory, destination path will retain input file's path as suffix,
+     * so destination path can't be matched with source path if archive directory's depth is
+     * longer than 2, as neither file nor parent directory of destination path can be matched
+     * with source path.
+     */
+    private val skipCheckingGlob: Boolean = baseArchivePath.exists(_.depth() > 2)
+
+    def archive(entry: FileEntry): Unit = {
+      require(baseArchivePath.isDefined)
+
+      if (sameFsSourceAndArchive) {
+        val curPath = new Path(new URI(entry.path))
+        val newPath = new Path(baseArchivePath.get, curPath.toUri.getPath.stripPrefix("/"))
+
+        if (!skipCheckingGlob && pathMatchesSourcePattern(newPath)) {
```

Comment:
If thought that you agreed with my suggestion that this should be done once during config validation?

Reply:
Not sure I understand. Could you elaborate? I agreed with your suggestion as it shouldn't be needed to be calculated per path and don't get anything except that.

If you meant about moving skipCheckingGlob into config validation, I'm not sure I like it, as it's just the optimization on FileStreamSourceCleaner, implementation details, so no need to expose it to outside.

Reply:
OK looking the comment again I only understood small amount of your comment. Sorry about this. The source path contains wildcard in any subpath so only depth check can be optimized. Please correct me if I'm missing here.

Reply:
My suggestion was that, during config validation, you should check whether a file that is moved under the archive directory can be detected by the source pattern. It shouldn't be hard to check that.

e.g. in your test:

pattern = /hello*/h{e,f}ll?
archive = /hello

I don't fully understand the matching you explained, but my guess is that you could end up with an archived file called "/hello/hello1/hell" and that somehow matches the pattern?

Anyway, that case is easy to check, because the archive path matches a subset of the pattern. Just yell at the user if the does that. They should pick a non-ambiguous path for the archive.

Reply:
> I don't fully understand the matching you explained, but my guess is that you could end up with an archived file called "/hello/hello1/hell" and that somehow matches the pattern?

Exactly. We cannot forecast if first depth of source path has wildcard. Yes we can also prune the case if the first depth of source path doesn't contain wildcard, but feels like more and more complicated.

As we would really want to simplify the case, how about just forcing baseArchivePath to have depth more than 2? (I feel it would be harmless in production - if they're encountering the case, just create subdirectory.) Then we can eliminate the pattern check entirely and throw error on config validation phase. WDYT?

Reply:
If you think doing the check is too hard, then sure, add that restriction. It's better than doing this check on every file being archived.

Reply:
Technically, we just need to check pattern only when the depth of base archive path is 2. The depth of base archive path cannot be 1, because then archive file will always refer to the same path of original source file. (Yes I was aware of this but forgot to apply the fact to prune.)

So `depth > 2` -> no need to check, `depth == 1` -> always matches, `depth == 2` -> need to check the pattern. We can even just simplify the case to avoid checking per file as checking the earliest dir name of base archive path matches the earlier dir name (including glob path) of source path. It won't guarantee all the archive paths will match the source path - that's why I did full pattern match per file, but if we really don't want to do pattern matching per file, yes, we can do it.

What I concern about is rather not a technical one. If we want to just yell and fail the query instead of skipping archive file, we must ensure end users understand why the query is failed and how to get it fixed. I had to explain most of reviewers how pattern check works, even with that I'm feeling reviewers don't feel familiar with it - I'm unsure it would work for end users though we can add some guidance how it works.

So we will need to simplify the condition as much as possible if we really want to just fail the query. The difference of necessary amount of explanation for depth > 2 and depth >=2 are significant, and the actual harm of not allowing depth == 2 is that they just need to create a subdirectory and use the subdirectory, which doesn't seem to be critical.

Does it make sense?

Reply:
The thing with users in cases like this is that you don't need to explain why. You can just say "your archive path must have a depth of at least 2 subdirectories" or something. It's a restriction of the feature, and why it exists is less important to users than the fact that it does.

Now, developers maintaining this code need to understand the why, which is why it's more important to keep things simple in the code.

Reply:
Also, as a user, I expect that if I enter an archiving configuration that will end up with none of my files being archives, that I'll see an error up front about that, not something hidden in log files.

Reply:
If I were one of end users and query fails because of glob path, I might be puzzling and want to know why - that's only when we allow `depth == 2` and I think we've agreed `depth > 2` doesn't need to explain further detail. I'll apply it.